[Tan Li En, Ryan] iP

ryantanlien opened the pull request:
# Butler
> "Your mind is for having idea, not holding them." - David Allen

Butler frees your mind of having to remember things you need to do. It's 
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/ryantanlien/ip)
2. double click it
3. add your tasks
4. let it manage your tasks for you :wink:

And it's **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

#

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

ryantanlien committed:
Butler Level 1

Level 1. Greet, Echo, Exit

ryantanlien committed:
Butler Level 2

Level 2. Add, List
Added the ability to store whatever text entered by the user and display them back to the user when requested.

ryantanlien committed:
Butler Level 3

Add the ability to mark tasks as done. Add the ability to change the status back to not done.

ryantanlien committed:
Butler Level 3.1

Incomplete Butler Level 4
Added 3 Task subtypes:
- Deadline class
- Event class
- ToDo class

Added toString method to Task class

ryantanlien committed:
Butler Level 4

Complete Butler Level 4
- Restructured Input Handling Blocks
- Completed toString() overridden methods for Deadline, Event and ToDo classes

ryantanlien committed:
Text UI Testing

Added Test Cases for Behavior for Deadlines , ToDos, and Events
- Creation (adding to list)
- Marking
- Unmarking
- Listing

ryantanlien committed:
Butler Level 5

Handled Errors
-Invalid Inputs for Deadline, ToDo, Events, such as empty description or empty date/time
-Invalid commands such as blah that have are not tied to Butler actions
-Invalid inputs for mark and unmark commands such as marking or unmarking an item in the list that does not exist, or following mark and unmark with multiple integers or non-integers

ryantanlien committed:
Butler Level 6

-Added try and catch block for NumberFormatExceptions in mark, delete and unmark sections of handleCommands

- Added testing and methods for deleting Tasks off the task ArrayList, catching the same exceptions as marking and unmarking Tasks in the list.

ryantanlien committed:
Butler Enums

Added the use of enums and switch statements in the handleCommand method for better refactorability in later iterations.

ryantanlien committed:
Add FileManager and resources folder

Lack of resources folder and I/O capability

Butler class functions as a Input and Output handler, need to separate File I/O responsibility from this terminal input and output class.

FileManager class is an incomplete File I/O handler whose main capability is to able to read and write to files.

ryantanlien committed:
Butler Level 7

Complete FileManager

- FileManager now outputs the list format in a file.
- If resources folder (or any parent folder) does not exist in main, then make the parent folders in the hierachy

ryantanlien committed:
Butler Level 7

Removed redundant test function call

ryantanlien committed:
Butler Level 8

Add LocalDateTimeManager
Edit Deadline and Event to be able to contain formatted LocalDateTimes as strings.

LocalDateTimeManager was created to separate area of responsibilities (processing strings to LocalDateTime formats) from the Event and Deadline classes

ryantanlien committed:
Merge branch 'branch-Level-7'
Merge branch-Level-7 to master branch

ryantanlien committed:
Merge branch 'branch-Level-8'
Merge branch-Level-8 with master branch

ryantanlien committed:
Refactor code structure

Remove Command Enum
Add Command Object Types
Create Parser, Ui and Storage classes

- The command enum was removed to bring in more OOP, as per the stretch goal.

- handleCommand method was split into multiple classes, Parser to parse inputs to commands, Ui to handle input and output, Storage to handle File I/O.

Note: Storage is a renaming of the previous FileManager

ryantanlien committed:
Organise classes into packages

- Move command related classes into command package
- Move task related classes into task package
- Move Parser, LocalDateTimeManager, Storage, Ui into util package

The classes are divided into packages for better access

ryantanlien committed:
Add JUnit 5 Tests

Add JUnit 5.6.0 dependency
Set up test module and rebuilt main module
Add simple ButlerTest
Add simple DeadlineTest

Misc:
Change access modifier on parser and Tasklist Methods to public to allow access to other classes in butler packages

ryantanlien committed:
Create MANIFEST.MF

Set-up project to be packaged as an executable JAR file via artifacts in IntelliJ IDEA.

ryantanlien committed:
Wrote Javadocs for butler sub-packages

Wrote header comments for public methods in butler sub-packages.

No header comments yet exist for Butler, ButlerInputException and TaskList classes

ryantanlien committed:
Update all code in Source root to match given style guide

Update code to match source guide.
Re-order import statements in alphabetical order.

ryantanlien committed:
Add find command functionality

Users can now find tasks in their list in the chatbot that contains the provided substring in their description.

New findCommand command object added to adhere to OOP principles.

ryantanlien committed:
Merge branch 'branch-A-JavaDoc'
Complete adding javadocs in separate branch.

ryantanlien committed:
Merge branch 'branch-A-CodingStandard'

ryantanlien committed:
Merge branch 'branch-Level-9'
Completed Level-9 in branch

ryantanlien committed:
Update Parser.java

Fixed merge bug where imports were changed from command.* to single imports, and FindCommand was left out of the imports.

ryantanlien committed:
Add Gradle support to project

Support Gradle in project

JUnit Dependencies were difficult to handle and module structure had to be defined manually. Test suite also had to be run manually to ensure no regressions occurred.

Gradle now takes care of:
-Compilation
-Running tests
-Installing and handling dependencies

This reduces the menial work to be done in the iP

Update .gitignore to ignore build.gradle file

ryantanlien committed:
Update .gitignore

Ignores .gradle files, specifically build.gradle

ryantanlien committed:
Merge branch 'branch-Gradle'
Support Gradle in project successfully

ryantanlien committed:
Support Checkstyle in project

Automate checking for code style with CheckStyle

Manual checking of code style is tedious. Automated checking is much more efficient.

Add checkstyle.xml and suppressions.xml
Add checkstyle checking functionality to build.gradle

redpelican2108 commented on src/main/java/butler/util/Parser.java:
```diff
@@ -0,0 +1,136 @@
+import butler.command.MarkCommand;
+import butler.command.TodoCommand;
+import butler.command.UnmarkCommand;
+import butler.command.FindCommand;
```
Shouldn't butler.command.FindCommand be Positioned after EventCommand and InvalidCommand?

redpelican2108 commented on src/main/java/butler/task/Event.java:
```diff
@@ -0,0 +1,47 @@
+
+public class Event extends Task {
+    
+    /* String representation of date and time not following any format */
```
Shouldn't Javadoc of class members use /** instead of /* ?

redpelican2108 commented on src/main/java/butler/command/FindCommand.java:
```diff
@@ -0,0 +1,19 @@
+    public FindCommand(String searchString) {
+        this.searchString = searchString;
+    }
+
```
Isn't Find Command missing it's javadoc?

redpelican2108 reviewed (commented):
I like how your code is very clean and easy to read. Keep up the good work

ryantanlien committed:
Design GUI with JavaFX

Text-base design was desired due to change in requirements, update to GUI

Refactor Ui class to become a handler class
Remove text-based I/O redirection testing
Add UiApp class as the driver for GUI window
Update command classes to store strings in Ui class instead of printing to output

ryantanlien committed:
Merge branch 'branch-Level-10'
Complete GUI

ryantanlien committed:
Update JavaDocs

Change JavaDocs to be compliant with the changes due to adding a GUI.

- Changed Command JavaDocs
- Added JavaDocs to FindCommand, and classes in Ui packages

ryantanlien committed:
Refactor Ui and UiApp

Split Responsibility of UiApp with Ui.

UiApp is responsible for launching and styling the GUI as well as for handling user inputs and changing the GUI dynamically according to input.

Ui now not only serves as a message holder but also to help the GUI handle inputs, and getting a response from Butler.

-Shifted handleInputs and static resources from UiApp to Ui

-Fixed a bug where inputting an unformatted date and time would cause the date and time to be added to the String representation of a task object

ryantanlien committed:
Add padding to chat text

Added padding to chat text and credited relevant sources for code reuse in GUI

justinekoh commented on src/main/java/butler/util/ui/DialogBox.java:
```diff
@@ -0,0 +1,49 @@
+    public DialogBox(Label l, ImageView iv) {
+        text = l;
+        displayPicture = iv;
+
```
Nice use of blank lines to separate logical code blocks! :)

joey-chance replied:
I agree! Your code is very readable. 👍  Good job and keep it up!

justinekoh commented on src/main/java/butler/util/Parser.java:
```diff
@@ -0,0 +1,137 @@
+        } else if (firstWord.equals("list") && stringArray.length == 1) {
+            return new ListCommand();
+        } else if (firstWord.equals("mark")) {
+            if (stringArray.length != 2) {
+                throw new ButlerInputException("Error. Sorry Master, " +
+                        "but a mark or unmark command must be followed by one Integer.\n");
+            }
+            try {
+                return new MarkCommand(Integer.parseInt(stringArray[1]) - 1);
+            } catch (NumberFormatException exception) {
+                throw new ButlerInputException("Error. Sorry Master, " +
+                       "but a mark or unmark command must be followed by one Integer.\n");
+            }
```
Is there a reason why you chose to handle the case where `stringArray.length != 2` explicitly instead of leaving it to the try-catch block?

ryantanlien replied:
Hi Justine! The reason I chose to handle the case of stringArray.length != 2 here is because it is the job of the parser  to check for input. 

In addition, the try catch block should be reserved for sections of code that throws exceptions further down the function call stack. Even if I did place the if statement in the try catch block, the parser would not know what to do with the exception message and would have to rethrow it anyway, so I just placed it outside the try catch block.

justinekoh reviewed (commented):
LGTM! Very nice structure and consistent formatting. :)

joey-chance commented on src/main/java/butler/TaskList.java:
```diff
@@ -0,0 +1,193 @@
+        }
+        return output;
+    }
+
```
I like how clean your code is. Good job and keep it up! 👍

joey-chance commented on src/main/java/butler/TaskList.java:
```diff
@@ -0,0 +1,193 @@
+     */
+    public String addDeadlineToList(String description, String dateTime) throws ButlerInputException {
+        if (description.equals("")) {
+            throw new ButlerInputException("Error. Sorry Master, but the description of a Deadline cannot be empty.\n");
```
Any reason why you did it this way instead of handling print statements in your Ui class?

ryantanlien replied:
Hi Joey! I did it this way because of Single Responsibility Principle, having the message in the exception allows me to handle it and print the message should the exception be thrown. Having the Ui handle all the print statements is probably not the best idea if we have to refactor our code (to pass the message to a GUI for example)

joey-chance reviewed (commented):
LGTM. Your code is very readable 👍

ryantanlien committed:
Change code style to fit checkstyle

Change sections of code in miscellaneous ways to fit code style.

Changing code style improves improves readability and code quality

ryantanlien committed:
Add Java Assertions to code in util package

Add assertions to util classes that handle program.

There are no explicit checks in the program to ensure that objects created by manager classes are properly instantiated.

There is no guarantee that during runtime unchecked exceptions such as NullPointerExceptions will not be thrown.

Java assertions were added in util package classes to make sure that objects and object parameters are not null and are properly created by their relevant classes.

This helps to ensure that in the future, bugs and unexpected exceptions during runtime is minimised.

ryantanlien committed:
Unified style to adhere to checkStyle

Fixed all styling errors with checkStyle

ryantanlien committed:
Refactor parse method

Shorten parse method

Parse method is getting very long as there are many keywords for commands to detect.

The long method results in poor code quality as reading the method is difficult, making maintenance difficult as well.

Extracting the different commands that we need to detect allows for better readability for readers as well as better maintainability, as it is easier to debug due to modularity.
This generally improves the quality of the code.

ryantanlien committed:
Merge pull request #1 from ryantanlien/branch-A-Assertions

Merge branch-A-Assertions

ryantanlien committed:
Merge branch 'master' into branch-A-CodeQuality

ryantanlien committed:
Merge pull request #4 from ryantanlien/branch-A-CodeQuality

Merge branch-A-CodeQuality

ryantanlien committed:
Add UpdateCommand

Added UpdateCommand functionality as part of C-Update feature.

Users can now update the descriptions of any tasks in the task list given their indexes. This can be done without deleting the original object.

The command format is "update (index of task to be updated) /description (updated description).

Made miscellaneous bug fixes
- "event" keyword would produce deadline instead
-  fixed testing where because code was updated since GUI commit, exceptions would be thrown resulting in failed tests.

Also made general changes to code to fit with style

ryantanlien committed:
Merge branch 'branch-BCD-Extension'
Merge branch-BCD-Extension into master

ryantanlien committed:
Add Butler greeting to GUI

ryantanlien committed:
Fix bye command functionality

ryantanlien committed:
Fix a bug that created a new line between each task when listing task list

ryantanlien committed:
Add chibi images for user and butler

ryantanlien committed:
Add functionality to load TaskList from storage file

Core functionality not implemented initially

Storage stores the entire list as a String in the storage file using their toString() methods. This makes it difficult to process when attempting to load, due to unnecessary newlines characters and unclean data.

Task members are now delimited by "|" character for easy loading. Loading functionality is also correctly added.

Miscellaneous Notes:
-Unify style with checkStyle
-Remove unused imports in Ui.java

ryantanlien committed:
Add Ui.png

ryantanlien committed:
Update README.md

Create user guide

ryantanlien committed:
Set theme jekyll-theme-tactile

ryantanlien committed:
Merge branch 'master' of https://github.com/ryantanlien/ip

ryantanlien committed:
Update list.txt

ryantanlien committed:
Update README.md

ryantanlien committed:
Update README.md headers

ryantanlien committed:
Add bye under usage to README.md

ryantanlien committed:
Update GUI

Add background to GUI and chatbox to text labels

The Ui is fairly plain and did not have enough personality. Adding a background and chatboxes simulates a scene and conversation between the user and Butler.

ryantanlien committed:
Update TaskList.java

Fixed a bug where keying in negative numbers or zero for delete, mark, unmark and update commands would throw an unchecked exception
